### PR TITLE
Use local complete block and local reduction block to identify compact dataflow

### PR DESCRIPTION
**Pull request by yzh119**
Per @Hzfengsy 's proposal: https://gist.github.com/Hzfengsy/eab83a98c0b034af1c6c7fac44bae741,
We use the concept of "local" complete block and "local" reduction block to identify whether a subtree has compact dataflow or not.
Compared to the original concept of complete block and reduction block, we only require a block to dominate a given subtree.

### CODE REVIEW

File: src/tir/schedule/analysis/analysis.cc [hunk 1]
```diff
@@ -145,23 +145,37 @@ ScopeBlockLoopInfo GetScopeBlockLoopInfo(const Block& scope_block) {
...
+ * \param scope_root_sref The StmtSRef corresponding to the root scope.
+ * \param block_sref The block whose dominant property is to be checked.
+ * \return A boolean indicating if the block is a dominant block.
  */
-bool IsDominantBlock(const BlockScope& scope, const StmtSRef& block_sref) {
+bool IsDominantBlock(const ScheduleState& self, const StmtSRef& scope_root_sref,
+                     const StmtSRef& block_sref) {
+  std::unordered_map<Buffer, Array<StmtSRef>, ObjectPtrHash, ObjectPtrEqual> buffer_writers;
+  const BlockNode* maybe_root_block = scope_root_sref->StmtAs<BlockNode>();
+  if (maybe_root_block) {
```
Comment (Hzfengsy) [inline src/tir/schedule/analysis/analysis.cc:159]: Please check `scope_root_sref` is inside or equal to the scope block if possible

File: src/tir/schedule/analysis/analysis.cc [hunk 2]
```diff
@@ -207,14 +220,14 @@ int CheckCompleteBlockErrorCode(const ScheduleState& self, const StmtSRef& block
 
 static const char* kCompleteBlockDefinition = R"(Definition of a complete block:
 1) All block vars are data parallel
-2) Dominant: the block is the only writer of its output, dominating the reader of its output buffers
+2) Dominant: the block is the only writer of its output, dominating the reader of its output buffers under the given scope.
```
Comment (Hzfengsy) [inline src/tir/schedule/analysis/analysis.cc:223]: for `kCompleteBlockDefinition`, I still prefer to keep it as `under the block scope`

File: src/tir/schedule/analysis/analysis.cc [hunk 3]
```diff
@@ -207,14 +220,14 @@ int CheckCompleteBlockErrorCode(const ScheduleState& self, const StmtSRef& block
...
 3) No overlap between the buffers the block reads and writes)";
 
 static const char* kReductionBlockDefinition = R"(Definition of a reduction block:
 1) The block has the `init` statement
 2) All the block bindings are quasi-affine expressions
 3) All block vars are either data parallel block vars or reduction block vars
-4) Dominant: the block is the only writer of its output, dominating the reader of its output buffers
+4) Dominant: the block is the only writer of its output, dominating the reader of its output buffers under the given scope.
```
Comment (Hzfengsy) [inline src/tir/schedule/analysis/analysis.cc:230]: +1

File: src/tir/schedule/analysis/analysis.cc [hunk 4]
```diff
@@ -375,12 +386,14 @@ void CheckSubtreeCompactDataflow(const ScheduleState& self, const StmtSRef& subt
     }
     String FastErrorString() const final {
       return "ScheduleError: The queried subtree root in SRef tree does not have compact dataflow, "
-             "because some of its child block on SRef tree is neither a complete block nor a "
+             "because some of its child block on SRef tree is neither a local complete block nor a "
+             "local "
```
Comment (Hzfengsy) [inline src/tir/schedule/analysis/analysis.cc:390]: It would be great if we can add the definition of `local complete`

**Review by Hzfengsy — CHANGES_REQUESTED** [review]

File: src/tir/schedule/analysis/analysis.cc (hunk 1 above)
Comment (yzh119) [reply src/tir/schedule/analysis/analysis.cc:159]: Done.

**Review by Hzfengsy — APPROVED** [review]

**Pull request merged**